src/preamble/02_what_is_a_module_and_a_package.md

```diff
@@ -1 +1,365 @@
+used by other packages, an **executable** produces a compiled binary that can be
+executed, a **test suite** is a program that can be invoked like an executable
+but it generally just tests code that is internal to the package.
+
+> Note: each of these components will receive individual treatment later on in
+> the guide. If you are interested to take a peek now you can find them here:
+> [library](../new_to_cabal/06_first_cabal_library.md),
+> [executable](../new_to_cabal/06_first_cabal_executable.md),
+> [test-suite](../leveling_up/02_first_cabal_test_suite.md)
+
+There are specific keys that we use to enumerate modules within components:
+`exposed-modules`, `other-modules`, `virtual-modules`, `test-module`, and even
+`main-is` which describes the entrypoint module for an executable. These fields
+are children of the top level component fields (mentioned above).
+
+External dependencies are declared at the component level, and reside under the
+key `build-depends` (a child of the various component fields). There are also
+foreign (non-Haskell) dependencies too which live under `foreign-library`, these
+are a bit anomalous in the sense that they are not associated with a single
+component, there is a separate
+[chapter](../getting_fancy/06_foreign_libraries.md) on them.
+
+It is not necessary to commit these fields to memory, but it is good to be aware
+of them. Hopefully when you look at the layout of a `.cabal` file next, you will
+see the `package <- component <- module` hierarchy reflected in the nested
+fields!
+
+## What is a package?
+
+If a module is the smallest unit of code in cabal, a **package** is the largest
+unit of distribution. At its core, all a package is is a distributable artefact
+that provides access to components.
+
+> You might remember cabal projects from earlier, as being _upstream_ of
+> packages. Good catch, a **project** is bigger in the sense that it is a
+> grouping of packages, but it is not distinct in terms of the distribution of
+> code.
+
+The `.cabal` file represents a single package, and is really just a collection
+of metadata about the packages constituent parts. It defines the packages
+components and their internal dependencies (list of modules) as well as the
+packages external dependencies; usually other packages.
+
+To make things more concrete, here is a pseudo-Haskell type representing the
+package-component-module hierarchy:
+
+```haskell
+data Module = Module { moduleName :: Text, moduleEntities :: [Entities] }
+
+data ComponentType = Executable | Library | TestSuite
+
+data Component =
+  Component
+    { componentName    :: Text
+    , componentType    :: ComponentType
+    , componentModules :: [Module]
+    }
+
+type Package = [Component]
+```
+
+> Note: technically a package can contain no components, but if you try and run
+> `cabal build` you will get an error telling you that your package contains no
+> components.
```

Comment:
perhaps replace "error telling you that your package contains no components" with something like "message suggesting the omission was erronous". Or "error assuming...".